File: umbra-typescript/tests/backup.test.ts
import { afterEach, describe, expect, it } from 'vitest'
import { gameBackup, MemoryDeviceCredentialStore, MemoryTokenStore, UmbraClient } from '../src'
import { json, readBody, startServer } from './helpers'

const servers: Array<{ close: () => Promise<void> }> = []

afterEach(async () => {
  await Promise.all(servers.splice(0).map(server => server.close()))
})

describe('backup client', () => {
  it('uploads a blob through presign PUT and confirm', async () => {
    let putBody = ''
    const object = await startServer(async (req, res) => {
      if (req.method === 'PUT' && req.url === '/object') {
        putBody = await readBody(req)
        res.writeHead(200)
        res.end()
        return
      }
      res.writeHead(404)
      res.end()
    })
    servers.push(object)

    const api = await startServer(async (req, res) => {
      if (req.method === 'POST' && req.url === '/api/v1/client/backup/presign') {
        const body = JSON.parse(await readBody(req)) as Record<string, unknown>
        expect(body).toMatchObject({
          category: 'game',
          subject: 'mc',
          version: 'v1',
          file_size: 5,
          content_type: 'text/plain',
        })
        json(res, 200, {
          code: 0,
          msg: 'success',
          data: {
            backup_id: 42,
            presigned_url: `${object.url}/object`,
            expires_in: 3600,
          },
        })
        return
      }
      if (req.method === 'POST' && req.url === '/api/v1/client/backup/confirm') {
        const body = JSON.parse(await readBody(req)) as Record<string, unknown>
        expect(body).toEqual({ backup_id: 42 })
        json(res, 200, {
          code: 0,
          msg: 'success',
          data: {
            backup_id: 42,
            size_bytes: 5,
            etag: 'etag',
            quota: { quota_bytes: 10, used_bytes: 5, available_bytes: 5 },
          },
        })
        return
      }
      json(res, 404, { code: 1001, msg: 'not found', data: null })
    })
    servers.push(api)

    const store = new MemoryTokenStore()
    await store.save({
      accessToken: 'token',
      tokenType: 'bearer',
      expiresAt: new Date(Date.now() + 3600_000).toISOString(),
    })
    const client = new UmbraClient({
      baseUrl: api.url,
      clientId: 'client',
      tokenStore: store,
      deviceStore: new MemoryDeviceCredentialStore({
        deviceId: 'dev_test',
        deviceSecret: 'device-secret',
      }),
    })

    const result = await client.backups.uploadBlob(
      gameBackup('mc', 'v1'),
      new Blob(['hello'], { type: 'text/plain' }),
    )

    expect(result.backupId).toBe(42)
    expect(result.sizeBytes).toBe(5)
    expect(putBody).toBe('hello')
  })
})

File: umbra-go/backup_test.go
package umbra

import (
	"context"
	"encoding/json"
	"io"
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"
	"time"
)

func TestUploadReaderPresignPutConfirm(t *testing.T) {
	var objectURL string
	var putBody string

	objectServer := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodPut {
			t.Fatalf("object method = %s", r.Method)
		}
		data, err := io.ReadAll(r.Body)
		if err != nil {
			t.Fatal(err)
		}
		putBody = string(data)
		w.WriteHeader(http.StatusOK)
	}))
	defer objectServer.Close()
	objectURL = objectServer.URL + "/object"

	mux := http.NewServeMux()
	mux.HandleFunc("/api/v1/client/backup/presign", func(w http.ResponseWriter, r *http.Request) {
		var req presignUploadRequest
		if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
			t.Fatal(err)
		}
		if req.Category != "game" || req.Subject != "mc" || req.Version != "v1" {
			t.Fatalf("request address = %+v", req)
		}
		_ = json.NewEncoder(w).Encode(map[string]any{
			"code": 0,
			"msg":  "success",
			"data": PresignUploadResult{BackupID: 42, PresignedURL: objectURL, ExpiresIn: 3600},
		})
	})
	mux.HandleFunc("/api/v1/client/backup/confirm", func(w http.ResponseWriter, r *http.Request) {
		var req backupTargetRequest
		if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
			t.Fatal(err)
		}
		if req.BackupID != 42 {
			t.Fatalf("backup_id = %d, want 42", req.BackupID)
		}
		_ = json.NewEncoder(w).Encode(map[string]any{
			"code": 0,
			"msg":  "success",
			"data": ConfirmUploadResult{
				BackupID:  42,
				SizeBytes: 5,
				ETag:      "etag",
				Quota:     QuotaInfo{QuotaBytes: 10, UsedBytes: 5, AvailableBytes: 5},
			},
		})
	})
	apiServer := httptest.NewServer(mux)
	defer apiServer.Close()

	store := NewMemoryTokenStore()
	if err := store.Save(context.Background(), &TokenSet{
		AccessToken: "token",
		TokenType:   "bearer",
		ExpiresAt:   time.Now().Add(time.Hour),
	}); err != nil {
		t.Fatal(err)
	}
	deviceStore := NewMemoryDeviceStore()
	if err := deviceStore.Save(context.Background(), &DeviceCredentials{DeviceID: "dev_test", DeviceSecret: "secret"}); err != nil {
		t.Fatal(err)
	}
	client, err := New(Config{
		BaseURL:     apiServer.URL,
		ClientID:    "client",
		RedirectURI: "http://127.0.0.1:1420/auth/callback",
		TokenStore:  store,
		DeviceStore: deviceStore,
	})
	if err != nil {
		t.Fatal(err)
	}

	result, err := client.Backup.UploadReader(
		context.Background(),
		GameBackup("mc", "v1"),
		strings.NewReader("hello"),
		5,
		UploadOptions{ContentType: "text/plain"},
	)
	if err != nil {
		t.Fatalf("UploadReader() error = %v", err)
	}
	if result.BackupID != 42 || result.SizeBytes != 5 {
		t.Fatalf("result = %+v", result)
	}
	if putBody != "hello" {
		t.Fatalf("putBody = %q, want hello", putBody)
	}
}
